quinn-proto/src/connection/pacing.rs

```diff
@@ -0,0 +1,125 @@
+//! Pacing of packet transmissions.
+
+use std::time::{Duration, Instant};
+
+use tracing::{debug, warn};
+
+/// A simple token-bucket pacer. The bucket starts full and has an adjustable capacity. Once the
+/// bucket is empty, further transmission is blocked. The bucket refills at a rate slightly faster
+/// than one contestion window per RTT.
+pub struct Pacer {
+    capacity: u64,
+    tokens: u64,
+    prev: Instant,
+}
+
+impl Pacer {
+    /// Obtains a new [`Pacer`].
+    pub fn new(capacity: u64, now: Instant) -> Self {
+        Self {
+            capacity,
+            tokens: capacity,
+            prev: now,
+        }
+    }
+
+    /// Record that a packet has been transmitted.
+    pub fn on_transmit(&mut self, packet_length: u16) {
+        self.tokens = self.tokens.saturating_sub(packet_length.into())
+    }
+
+    /// Return how long we need to wait before sending a packet.
+    ///
+    /// If we can send a packet right away, this returns `None`. Otherwise, returns `Some(d)`,
+    /// where `d` is the time before this function should be called again.
+    ///
+    /// The 5/4 ratio used here comes from the suggestion that N = 1.25 in the draft IETF RFC for
+    /// QUIC.
+    pub fn delay(
+        &mut self,
+        smoothed_rtt: Duration,
+        mtu: u16,
+        window: u64,
+        now: Instant,
+    ) -> Option<Instant> {
+        debug_assert_ne!(
+            window, 0,
+            "zero-sized congestion control window is nonsense"
+        );
+
+        // if we can already send a packet, there is no need for delay
+        if self.tokens > mtu.into() {
+            return None;
+        }
+
+        // we disable pacing for extremely large windows
+        if window > u32::max_value().into() {
+            debug!("window too large, disabling pacing");
+            return None;
+        }
+
+        let time_elapsed = now.checked_duration_since(self.prev).unwrap_or_else(|| {
+            warn!("received a timestamp early than a previous recorded time, ignoring");
+            Default::default()
+        });
+
+        if smoothed_rtt.as_nanos() == 0 {
+            return None;
+        }
+
+        let elapsed_rtts = time_elapsed.as_secs_f64() / smoothed_rtt.as_secs_f64();
+        let new_tokens = window as f64 * 1.25 * elapsed_rtts;
+        self.tokens = self
+            .tokens
+            .saturating_add(new_tokens as _)
+            .min(self.capacity);
+
+        // if we can already send a packet, there is no need for delay
+        if self.tokens > mtu.into() {
+            return None;
+        }
+
+        let rtt = smoothed_rtt
```

Comment:
`srtt * mtu` is definitely not an rtt. I'm not sure what it *is* exactly, but we shouldn't call it RTT.

Reply:
`time_to_transmit_mtu` perhaps?

Reply:
The time required to transmit an MTU depends mainly on the congestion window, which isn't involved in this specific computation at all. `rtt_times_mtu` would be acceptable, if nothing else.

Reply:
`rtt_times_mtu` would be pretty meh IMO. Can't we come up with any name that reflects what this value represents semantically? If not, maybe we should move around some of the arithmetic operations so that any named values are semantically comprehensible?

Reply:
If I move the division up, it could be `unscaled_delay`.

Reply:
Sounds better to me!